pkg/deps/resolve.go

```diff
@@ -0,0 +1,167 @@
+	componentVersions map[string]*sortedVersions
+	metaLookup        map[bundle.ComponentReference]*depMeta
+	componentLookup   map[bundle.ComponentReference]*bundle.Component
+}
+
+// NewResolver takes in a list of components that make up the universe.
+func NewResolver(components []*bundle.Component) (*Resolver, error) {
+	svm, lup, err := sortedMapFromComponents(components)
+	if err != nil {
+		return nil, err
+	}
+	componentLookup := make(map[bundle.ComponentReference]*bundle.Component)
+	for _, c := range components {
+		componentLookup[c.ComponentReference()] = c
+	}
+
+	return &Resolver{
+		componentVersions: svm,
+		metaLookup:        lup,
+		componentLookup:   componentLookup,
+	}, nil
+}
+
+// Component retrieves a component (which is copied for safety), returning nil
+// if isn't known by the resolver.
+func (r *Resolver) Component(ref bundle.ComponentReference) *bundle.Component {
+	c := r.componentLookup[ref]
+	if c == nil {
+		return nil
+	}
+	return c.DeepCopy()
+}
+
+// HasComponent indicates whether a component is known by the resolver.
+func (r *Resolver) HasComponent(ref bundle.ComponentReference) bool {
+	return r.componentLookup[ref] != nil
+}
+
+// Add adds components to the resolver, returning a new resolver or
+// an error if it fails. If one of the new components provided has the same
+// name/version, it overwrites the existing component map.
+func (r *Resolver) Add(newComps []*bundle.Component) (*Resolver, error) {
+	m := make(map[bundle.ComponentReference]*bundle.Component)
+	for k, c := range r.componentLookup {
+		m[k] = c
+	}
+	for _, c := range newComps {
+		m[c.ComponentReference()] = c
+	}
+	var cs []*bundle.Component
+	for _, c := range m {
+		cs = append(cs, c)
+	}
+	return NewResolver(cs)
+}
+
+// ComponentVersions gets the versions for a particular component
+func (r *Resolver) ComponentVersions(comp string) ([]bundle.ComponentReference, error) {
+	cv, ok := r.componentVersions[comp]
+	if !ok {
+		return nil, fmt.Errorf("unknown component %q", comp)
+	}
+	var versions []bundle.ComponentReference
+	for _, ver := range cv.versions {
+		versions = append(versions, bundle.ComponentReference{
+			ComponentName: cv.componentName,
+			Version:       ver.version.String(),
+		})
+	}
+	return versions, nil
+}
+
+// ResolveOptions are options for resolving dependencies
+type ResolveOptions struct {
+	// Match are component annotations that must to be present. This is useful
+	// for matching positive characteristics of a component (example: this
+	// component passed qualification). Only one of the list of values need
+	// be present.
+	//
+	// In other words, this is a logical AND operation of the passed in Annotation and
+	// the component Annotation.
+	Match map[string][]string
+
+	// MatchIfPresent are component annotations that are analyzed if they are
+	// present on the component. This is useful for synchronizing values between
+	// the universe of components and the caller.
+	MatchIfPresent map[string][]string
```

Comment:
`MatchIfPresent` could use an example in the description of matching with and without present annotation